### PR TITLE
Feature to fetch authentication info from context when auth is enabled

### LINKED ISSUE #1094: Feature to fetch claims when authentication is enabled.
**Is your feature request related to a problem? Please describe.**
We can have a feature to get claims from GoFr context ....no matter the kind of authentication we use. 

**Describe the solution you'd like**
The usecase can be generalized for various kind of authentication like jwt or api key etc .....

### PR DESCRIPTION
**Pull request by coolwednesday**
## Pull Request Template

**Description:**

- Added a method GetAuthInfo() in Request Interface to fetch Auth Info when authentication is enabled.
- Closes #1094 

**Checklist:**

-   [x] I have formatted my code using  `goimport`  and  `golangci-lint`.
-   [x] All new code is covered by unit tests.
-   [x] This PR does not decrease the overall code coverage.
-   [x] I have reviewed the code comments and documentation for clarity.

**Thank you for your contribution!**

### CODE REVIEW

File: pkg/gofr/http/middleware/oauth.go [hunk 1]
```diff
@@ -26,6 +26,8 @@ type authMethod int
 
 const (
 	JWTClaim authMethod = iota // JWTClaim represents the key used to store JWT claims within the request context.
+	Username authMethod = 1    // Username fetches the username when we use Basic Auth
```
Comment (vipul-rawat) [inline pkg/gofr/http/middleware/oauth.go:29]: These constants are not part of the OAuth, please declare them where they are used.

Comment (ccoVeille) [reply pkg/gofr/http/middleware/oauth.go:29]: Also it's iota starting at 1, so the constant would have same value as previous variable?

File: pkg/gofr/http/request.go [hunk 2]
```diff
@@ -11,6 +11,9 @@ import (
 	"reflect"
 	"strings"
 
+	"github.com/golang-jwt/jwt/v5"
+	"gofr.dev/pkg/gofr/http/middleware"
+
 	"github.com/gorilla/mux"
```
Comment (ccoVeille) [inline pkg/gofr/http/request.go:17]:
```suggestion
	"github.com/golang-jwt/jwt/v5"
	"github.com/gorilla/mux"

	"gofr.dev/pkg/gofr/http/middleware"
```

File: pkg/gofr/http/request.go [hunk 3]
```diff
@@ -157,3 +160,21 @@ func (r *Request) bindForm(ptr any, isMultipart bool) error {
...
+	if r.Context().Value(middleware.JWTClaim) != nil {
+		m["JWTClaims"] = r.Context().Value(middleware.JWTClaim).(jwt.MapClaims)
+	}
+
+	if r.Context().Value(middleware.Username) != nil {
+		m["Username"] = r.Context().Value(middleware.Username).(string)
+	}
+
+	if r.Context().Value(middleware.APIKey) != nil {
+		m["ApiKey"] = r.Context().Value(middleware.APIKey).(string)
```
Comment (ccoVeille) [inline pkg/gofr/http/request.go:176]:
```suggestion
		m["APIKey"] = r.Context().Value(middleware.APIKey).(string)
```

File: pkg/gofr/http/request_test.go [hunk 4]
```diff
@@ -12,6 +12,9 @@ import (
 	"strings"
 	"testing"
 
+	"github.com/golang-jwt/jwt/v5"
+	"gofr.dev/pkg/gofr/http/middleware"
+
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
```
Comment (ccoVeille) [inline pkg/gofr/http/request_test.go:20]:
```suggestion
	"github.com/golang-jwt/jwt/v5"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"

	"gofr.dev/pkg/gofr/http/middleware"

```

File: pkg/gofr/http/request_test.go [hunk 5]
```diff
@@ -272,3 +275,45 @@ func TestBind_FormURLEncoded(t *testing.T) {
...
+func TestGetAuthInfo_ApiKey(t *testing.T) {
+	req := httptest.NewRequest(http.MethodGet, "/", http.NoBody)
+
+	ctx := context.WithValue(req.Context(), middleware.APIKey, "9221e451-451f-4cd6-a23d-2b2d3adea9cf")
+	*req = *req.Clone(ctx)
+	gofrRq := NewRequest(req)
+
+	u := gofrRq.GetAuthInfo()
+
+	assert.Equal(t, "9221e451-451f-4cd6-a23d-2b2d3adea9cf", u["ApiKey"])
```
Comment (ccoVeille) [inline pkg/gofr/http/request_test.go:300]:
```suggestion

```suggestion
	assert.Contains(t, u, "ApiKey")
	assert.Equal(t, "9221e451-451f-4cd6-a23d-2b2d3adea9cf", u["ApiKey"])
```

File: pkg/gofr/http/request_test.go [hunk 6]
```diff
@@ -272,3 +275,45 @@ func TestBind_FormURLEncoded(t *testing.T) {
...
+
+	req := httptest.NewRequest(http.MethodGet, "/", http.NoBody)
+
+	ctx := context.WithValue(req.Context(), middleware.JWTClaim, claims)
+	*req = *req.Clone(ctx)
+	gofrRq := NewRequest(req)
+
+	u := gofrRq.GetAuthInfo()
+
+	assert.Equal(t, claims, u["JWTClaims"])
```
Comment (ccoVeille) [inline pkg/gofr/http/request_test.go:318]: Same add a contains, it helps debugging

File: pkg/gofr/http/request.go [hunk 7]
```diff
@@ -157,3 +160,21 @@ func (r *Request) bindForm(ptr any, isMultipart bool) error {
...
+
+func (r *Request) GetAuthInfo() map[string]any {
+	m := make(map[string]any, 0)
+
+	if r.Context().Value(middleware.JWTClaim) != nil {
+		m["JWTClaims"] = r.Context().Value(middleware.JWTClaim).(jwt.MapClaims)
+	}
+
+	if r.Context().Value(middleware.Username) != nil {
+		m["Username"] = r.Context().Value(middleware.Username).(string)
```
Comment (ccoVeille) [inline pkg/gofr/http/request.go:172]: I would use UserName (uppercase on Name) to use camelCase, here and everywhere

**Comment by reski-rukmantiyo** [comment]
Hi @coolwednesday, do you have any complete examples for this? Thanks

**Comment by ccoVeille** [comment]
@coolwednesday why was the PR closed and branch deleted?

**Comment by ccoVeille** [comment]
Make sense thanks.

It's always a good thing to drop a message when closing a PR.

Also to crossref PR/issues

**Comment by reski-rukmantiyo** [comment]
> > Hi @coolwednesday, do you have any complete examples for this? Thanks
> 
> Hey @reski-rukmantiyo , you can see the release notes , of release [v1.25.0](https://github.com/gofr-dev/gofr/releases/tag/v1.25.0). In case of any doubts, feel free to ping us.
> 
> However, the issue mentioned here has been resolved in PR #1161 , for your reference.

Thanks @coolwednesday

**Pull request closed without merging**